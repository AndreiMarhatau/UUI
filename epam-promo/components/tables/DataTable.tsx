import * as React from 'react';
import {
    applyColumnsConfig, ColumnsConfig, DataRowProps, DataSourceState, Lens, ScrollManager, UuiContexts, getColumnsConfig, DataColumnProps, IEditable, DataTableState, DataSourceListProps, DataTableColumnsConfigOptions, UuiContext,
} from '@epam/uui';
import { IScrollbarsPositionValues } from '@epam/uui-components';
import { ColumnsConfigurationModal, DataTableHeaderRow, DataTableRow, DataTableScrollRow, DataTableMods } from './';
import { FlexRow, VirtualList } from '../';
import * as css from './DataTable.scss';

export interface DataTableProps<TItem, TId> extends IEditable<DataTableState>, DataSourceListProps, DataTableColumnsConfigOptions {
    getRows(): DataRowProps<TItem, TId>[];
    columns: DataColumnProps<TItem, TId>[];
    renderRow?(props: DataRowProps<TItem, TId>): React.ReactNode;
    renderNoResultsBlock?(): React.ReactNode;
    onScroll?(value: IScrollbarsPositionValues): void;
    showColumnsConfig?: boolean;
}

export class DataTable<TItem, TId = any> extends React.Component<DataTableProps<TItem, TId> & DataTableMods, any> {
    static contextType = UuiContext;
    context: UuiContexts;

    scrollManager = new ScrollManager();
    lens = Lens.onEditableComponent<DataSourceState>(this);

    getColumns() {
        return applyColumnsConfig(this.props.columns, this.getColumnsConfig());
    }

    setColumnsConfig = (config: ColumnsConfig) => {
        this.props.onValueChange({ ...this.props.value, columnsConfig: config });
    }

    getColumnsConfig() {
        return getColumnsConfig(this.props.columns, this.props.value.columnsConfig);
    }

    getDefaultColumnsConfig() {
        return getColumnsConfig(this.props.columns, {});
    }

    getRows() {
        const renderRow = this.props.renderRow || this.renderRow;

        return this.props.getRows()
            .map((row: DataRowProps<TItem, TId>) => renderRow({ ...row, scrollManager: this.scrollManager, columns: this.getColumns() }));
    }

    renderRow = (props: DataRowProps<TItem, TId>) => (
        <DataTableRow
            key={ props.rowKey }
            size={ this.props.size }
            borderBottom={ this.props.border }
            { ...props }
        />
    )

    onConfigurationButtonClick = () => {
        this.context.uuiModals.show<ColumnsConfig>(modalProps => (
                <ColumnsConfigurationModal
                    { ...modalProps }
                    columns={ this.props.columns }
                    columnsConfig={ this.getColumnsConfig() }
                    defaultConfig={ this.getDefaultColumnsConfig() }
                />
            ))
            .then(this.setColumnsConfig)
            .catch(() => null);
    }

    render() {
        return (
            <>
                <DataTableHeaderRow
                    key='header'
                    scrollManager={ this.scrollManager }
                    columns={ this.getColumns() }
                    onConfigButtonClick={ this.props.showColumnsConfig && this.onConfigurationButtonClick }
                    selectAll={ this.props.selectAll }
                    size={ this.props.size }
                    textCase={ this.props.headerTextCase }
                    allowColumnsReordering={ this.props.allowColumnsReordering }
                    allowColumnsResizing={ this.props.allowColumnsResizing }
                    { ...this.lens.toProps() }
                />
                <FlexRow key='body' topShadow background='white' cx={ css.body }>
                    <VirtualList
                        { ...this.lens.toProps() }
                        onScroll={ this.props.onScroll }
                        rows={ this.getRows() }
                        rowsCount={ this.props.rowsCount }
                        focusedIndex={ this.props.value?.focusedIndex }
                        shadow='dark'
                    />
                </FlexRow>
                <DataTableScrollRow key='scroll' scrollManager={ this.scrollManager } columns={ this.getColumns() }/>
            </>
        );
    }
}
