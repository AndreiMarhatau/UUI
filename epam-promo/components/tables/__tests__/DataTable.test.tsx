import React from 'react';
import { DataTable } from '../DataTable';
import renderer from 'react-test-renderer';

jest.mock('react-dom', () => ({
    findDOMNode: jest.fn(),
}));

class ResizeObserverMock {
    observe = () => jest.fn();
    unobserve = () => jest.fn();
    disconnect = () => jest.fn();
}

global.ResizeObserver = ResizeObserverMock;

describe('DataTable', () => {
    it('should be rendered correctly', () => {
        const tree = renderer
            .create(<DataTable
                columns={ [] }
                getRows={ () => [] }
                value={ {} }
                onValueChange={ jest.fn }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});


