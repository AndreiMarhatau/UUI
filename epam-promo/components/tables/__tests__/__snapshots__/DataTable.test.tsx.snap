// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should be rendered correctly 1`] = `
Array [
  <div
    className="container uui-table-row-container root uui-table-header-row container align-items-top"
  >
    <div
      className="scrollableColumnsWrapper"
      style={
        Object {
          "flexGrow": 100,
          "flexShrink": 1,
          "minWidth": 0,
          "width": 0,
        }
      }
    >
      <div
        className="scrollableColumnsContainer"
      >
        <div
          className="container align-items-top"
        />
      </div>
      <div
        className="scrollShadowLeft"
      />
      <div
        className="scrollShadowRight"
      />
    </div>
  </div>,
  <div
    className="root size-36 background-white topShadow body container align-items-center"
  >
    <div
      className="wrapper root shadow-dark"
    >
      <div
        className="os-host"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <div
          className="os-resize-observer-host"
        />
        <div
          className="os-padding"
        >
          <div
            className="os-viewport"
          >
            <div
              className="os-content"
            >
              <div
                className="container2"
                style={
                  Object {
                    "minHeight": null,
                  }
                }
              >
                <div
                  className="container3"
                  style={
                    Object {
                      "marginTop": 0,
                    }
                  }
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="os-scrollbar os-scrollbar-horizontal "
        >
          <div
            className="os-scrollbar-track"
          >
            <div
              className="os-scrollbar-handle"
            />
          </div>
        </div>
        <div
          className="os-scrollbar os-scrollbar-vertical"
        >
          <div
            className="os-scrollbar-track"
          >
            <div
              className="os-scrollbar-handle"
            />
          </div>
        </div>
        <div
          className="os-scrollbar-corner"
        />
      </div>
      <div
        className="uui-scroll-shadow-top"
      />
      <div
        className="uui-scroll-shadow-bottom"
      />
    </div>
  </div>,
  <div
    className="uui-table-scroll-row-container cellClass"
  >
    <div
      className="container uui-table-row-container uui-table-scroll-row container align-items-top"
    >
      <div
        className="os-host"
        style={
          Object {
            "flexGrow": 100,
            "flexShrink": 1,
            "minWidth": 0,
            "width": 0,
          }
        }
      >
        <div
          className="os-resize-observer-host"
        />
        <div
          className="os-padding"
        >
          <div
            className="os-viewport"
          >
            <div
              className="os-content"
            >
              <div
                className="container align-items-top"
              />
            </div>
          </div>
        </div>
        <div
          className="os-scrollbar os-scrollbar-horizontal "
        >
          <div
            className="os-scrollbar-track"
          >
            <div
              className="os-scrollbar-handle"
            />
          </div>
        </div>
        <div
          className="os-scrollbar os-scrollbar-vertical"
        >
          <div
            className="os-scrollbar-track"
          >
            <div
              className="os-scrollbar-handle"
            />
          </div>
        </div>
        <div
          className="os-scrollbar-corner"
        />
      </div>
    </div>
  </div>,
]
`;
